R/joins.R

```diff
@@ -31,49 +31,66 @@
-    }
-    y <- dplyr::auto_copy(x, y, copy = copy)
-
-    x <- data.table::copy(x)
-    y <- data.table::copy(y)
-    data.table::setkeyv(x, by$x)
-    data.table::setkeyv(y, by$x)
-    out <- op
-    grouped_dt(out, groups(x))
-  })
-
-  f <- eval(template, parent.frame())
-  attr(f, "srcref") <- NULL # fix so prints correctly
-  f
-  # nocov end
+#' @rdname join.tbl_dt
+inner_join.data.table <- function(x, y, by = NULL, copy = FALSE, ...){
+  by <- dplyr::common_by(by, x, y)
+  y <- dplyr::auto_copy(x, y, copy = copy)
+  out <- merge(x, y, by.x = by$x, by.y = by$y, all = FALSE, allow.cartesian = TRUE)
+  grouped_dt(out, groups(x)) 
 }
 
 #' @rdname join.tbl_dt
-inner_join.data.table <- join_dt({merge(x, y, by = by$x, allow.cartesian = TRUE)})
+left_join.data.table <- function(x, y, by = NULL, copy = FALSE, ...){
+  by <- dplyr::common_by(by, x, y)
+  y <- dplyr::auto_copy(x, y, copy = copy)
+  out <- merge(x, y, by.x = by$x, by.y = by$y, all.x = TRUE, allow.cartesian = TRUE)
+  grouped_dt(out, groups(x)) 
+}
 
 #' @rdname join.tbl_dt
-left_join.data.table  <- join_dt({merge(x, y, by = by$x, all.x = TRUE, allow.cartesian = TRUE)})
+right_join.data.table <- function(x, y, by = NULL, copy = FALSE, ...){
+  by <- dplyr::common_by(by, x, y)
+  y <- dplyr::auto_copy(x, y, copy = copy)
+  out <- merge(x, y, by.x = by$x, by.y = by$y, all.y = TRUE, allow.cartesian = TRUE)
+  grouped_dt(out, groups(x)) 
+}
 
 #' @rdname join.tbl_dt
-right_join.data.table  <- join_dt(merge(x, y, by = by$x, all.y = TRUE, allow.cartesian = TRUE))
+full_join.data.table <- function(x, y, by = NULL, copy = FALSE, ...){
+  by <- dplyr::common_by(by, x, y)
+  y <- dplyr::auto_copy(x, y, copy = copy)
+  out <- merge(x, y, by.x = by$x, by.y = by$y, all = TRUE, allow.cartesian = TRUE)
+  grouped_dt(out, groups(x)) 
+}
 
-#' @rdname join.tbl_dt
-semi_join.data.table  <- join_dt({
-  # http://stackoverflow.com/questions/18969420/perform-a-semi-join-with-data-table
-  w <- unique(x[y, which = TRUE, allow.cartesian = TRUE])
-  w <- w[!is.na(w)]
-  x[w]
-})
+
+trim_y_for_semi_join <- function(y, by){
+  by_x <- by$x
+  by_y <- by$y
+  y_trimmed <- y[, by_y, with = FALSE]
+  names(y_trimmed) <- by_x
+  y_trimmed
+}
 
 #' @rdname join.tbl_dt
-anti_join.data.table <- join_dt({x[!y, allow.cartesian = TRUE]})
+semi_join.data.table <- function(x, y, by = NULL, copy = FALSE, ...) {
+  by <- dplyr::common_by(by, x, y)
+  y <- dplyr::auto_copy(x, y, copy = copy)
+  by_x <- by$x
+  y_trimmed <- trim_y_for_semi_join(y, by)
+  w <- x[y_trimmed, which = TRUE, on = by_x, nomatch = 0L]
+  out <- x[sort(unique(w))]
+  grouped_dt(out, groups(x))
+}
 
 #' @rdname join.tbl_dt
-# http://stackoverflow.com/a/15170956/946850
-full_join.data.table <- join_dt({merge(x, y, by = by$x, all = TRUE, allow.cartesian = TRUE)})
+anti_join.data.table <- function(x, y, by = NULL, copy = FALSE, ...) {
+  by <- dplyr::common_by(by, x, y)
+  y <- dplyr::auto_copy(x, y, copy = copy)
+  by_x <- by$x
+  by_x <- by$x
```

Comment:
Can you please double-check, also if we could perhaps write `by$x` instead? The docs for the `on` argument suggest that we could also avoid renaming the columns in `y` altogether:

```r
on <- rlang::set_names(by$y, by$x)
w <- x[!y, which = TRUE, on = on]
...
```

Would that work with data.table 1.9.6?

dplyr already imports rlang, so no extra dependency added, but we need to explicitly declare the import in DESCRIPTION to be able to use it here.